src/ast.rs

```diff
@@ -1705,7 +1705,10 @@ impl Parse for Name {
                 let idx = subs.insert(Substitutable::UnscopedTemplateName(name));
                 let handle = UnscopedTemplateNameHandle::BackReference(idx);
 
-                let (args, tail) = TemplateArgs::parse(ctx, subs, tail)?;
+                let (args, tail) = TemplateArgs::parse(ctx, subs, tail).map_err(|err| {
+                    subs.pop();
```

Comment:
I looked around and this is the only spot I found where we insert something into the table and then do an operation that can immediately fail. I still think it's possible to have a situation as part of parsing a complex structure we create table entries and then the parsing operation later fails. Not sure what to do about that.